module/zfs/dmu_object.c

```diff
@@ -50,6 +59,9 @@ dmu_object_alloc_dnsize(objset_t *os, dmu_object_type_t ot, int blocksize,
 	dnode_t *dn = NULL;
 	int dn_slots = dnodesize >> DNODE_SHIFT;
 	boolean_t restarted = B_FALSE;
+	uint64_t *cpuobj = &os->os_obj_next_array[CPU_SEQID %
+	    os->os_obj_next_array_len];
+	int dnodes_per_chunk = 1 << dmu_object_alloc_chunk_shift;
```

Comment:
I'm not exactly sure what the situation is on Illumos.  But under Linux `CPU_SEQID % os->os_obj_next_array_len]` won't guarantee that there aren't multiple CPUs concurrently in this function operating on the same `cpuobj`.  I'm assuming the same it true for Illumos because between the atomic updates of `cpuobj` and the taking of `dn->dn_struct_rwlock` lock this looks to be handled nicely.

For the record this could happen as result on CPU hot plugging when additional processors come online and `os->os_obj_next_array` isn't resized.  Or more likely the process is moved to a different CPU during `schedule()`.   Or potentially it's rescheduled at any time when using a real-time pre-emptible kernel build.